kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/storage/FileBasedKieServerTemplateStorage.java

```diff
@@ -0,0 +1,169 @@
+    public static final String SERVER_TEMPLATE_FILE_NAME_PROP = "org.kie.server.controller.templatefile";
+    public static final String DEFAULT_SERVER_TEMPLATE_FILENAME = System.getProperty("java.io.tmpdir")+
+    		System.getProperty("file.separator")
+    		+"template_store.xml";
+    private static Logger logger = LoggerFactory.getLogger(FileBasedKieServerTemplateStorage.class);
+    private Map<String, ServerTemplate> templateMap = new ConcurrentHashMap<>();
+    private Map<String, ServerTemplateKey> templateKeyMap = new ConcurrentHashMap<>();
+    private String templatesLocation = System.getProperty(SERVER_TEMPLATE_FILE_NAME_PROP, DEFAULT_SERVER_TEMPLATE_FILENAME);
+    private Marshaller templateMarshaller = MarshallerFactory.getMarshaller(MarshallingFormat.XSTREAM,ServerTemplate.class.getClassLoader());
+
+
+    /**
+     * Writes the map of server templates to the file pointed at by templatesLocation
+     */
+    private synchronized void writeTemplateMap() {
+        try (FileWriter writer = new FileWriter(templatesLocation)) {
+            ((XStreamMarshaller)templateMarshaller).getXstream()
+            	.toXML(new ArrayList<ServerTemplate>(templateMap.values()), writer);
+        } catch (Throwable e) {
+            logger.error("Unable to write template maps for standalone controller",e);
+        }
+    }
+
+    /**
+     * Loads the map of server templates from the file pointed at by the templatesLocation
+     */
+    @SuppressWarnings("unchecked")
+    private synchronized void loadTemplateMapsFromFile() {
+        ArrayList<ServerTemplate> templates = null;
+        try (FileReader reader = new FileReader(templatesLocation)) {
+            templates = (ArrayList<ServerTemplate>)((XStreamMarshaller)templateMarshaller).getXstream().fromXML(reader);
+        } catch (Throwable e) {
+            logger.error("Unable to read server template maps from file",e);
+        }
+        if (templates != null && !templates.isEmpty()) {
+            templates.forEach(template -> {
+                templateKeyMap.put(template.getId(),new ServerTemplateKey(template.getId(),template.getName()));
+                templateMap.put(template.getId(),template);
+            });
+        }
+    }
+
+
+    @Override
+    public ServerTemplate store(ServerTemplate serverTemplate) {
+        ServerTemplate template = null;
+        synchronized (templateMap) {
+            templateKeyMap.put(serverTemplate.getId(),
+                               new ServerTemplateKey(serverTemplate.getId(),
+                                                     serverTemplate.getName()));
+            template = templateMap.put(serverTemplate.getId(),
+                                   serverTemplate);
+            writeTemplateMap();
+        }
+        return template;
+    }
+
+    @Override
+    public List<ServerTemplateKey> loadKeys() {
+        if (templateKeyMap.isEmpty()) {
+            loadTemplateMapsFromFile();
```

Comment:
I'd say this should be loaded when new instance of this class is being initialized otherwise there is a chance that multiple loads will be performed in multi threaded environment - e.g. when several kie servers connects at pretty much same time.